File: python/scripts/week-06/demo-03-conversational-agent.py
"""
Demo 3: Conversational Agent — talk to DevBuddy

Type queries. The dynamic agent plans, calls tools, and responds.
Type 'quit' to exit, 'cost' to see cumulative spending.

Run: python scripts/week-06/demo-03-conversational-agent.py
"""
import os, sys
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from src.agent import run_dynamic_agent

print("=" * 70)
print("  DevBuddy Agent — Conversational Mode")
print("=" * 70)
print()
print("  I can check build status, deployment history, active")
print("  incidents, and retrieve documentation. Ask me anything")
print("  about auth-service, payment-api, or inventory-service.")
print()
print("  Commands:  quit  |  cost  |  help")
print()

total_cost = 0.0
turn = 0

while True:
    try:
        query = input("  You → ").strip()
    except (EOFError, KeyboardInterrupt):
        break

    if not query:
        continue
    if query.lower() in ("quit", "exit", "q"):
        break
    if query.lower() == "cost":
        print(f"  💰 Total cost this session: ~${total_cost:.6f}")
        print()
        continue
    if query.lower() == "help":
        print("  Try: 'Is payment-api healthy?'")
        print("       'What was deployed recently for auth-service?'")
        print("       'Any incidents for inventory-service?'")
        print("       'Give me a full readiness report for payment-api'")
        print()
        continue

    turn += 1
    print(f"  ⏳ Turn {turn} — planning and executing...")
    result = run_dynamic_agent(query)
    total_cost += result.get("cost", 0)

    print(f"  Steps: {result['steps']}  |  Cost: ~${result['cost']:.6f}")
    print(f"  ─────────────────────────────────────────────")
    print(f"  {result['report']}")
    print()

print()
print(f"  Session ended. {turn} queries, total cost ~${total_cost:.6f}")
print("=" * 70)

File: nodejs/scripts/week-06/demo-03-conversational-agent.js
/**
 * Demo 3: Conversational Agent — talk to DevBuddy
 *
 * Type queries. The dynamic agent plans, calls tools, and responds.
 * Type 'quit' to exit, 'cost' to see cumulative spending.
 *
 * Requires: MCP server running (node src/mcp_server.js in another terminal)
 * Run: node scripts/week-06/demo-03-conversational-agent.js
 */
import { createInterface } from "readline";
import { runDynamicAgent } from "../../src/agent.js";

console.log("=".repeat(70));
console.log("  DevBuddy Agent — Conversational Mode");
console.log("=".repeat(70));
console.log();
console.log("  I can check build status, deployment history, active");
console.log("  incidents, and retrieve documentation. Ask me anything");
console.log("  about auth-service, payment-api, or inventory-service.");
console.log();
console.log("  Commands:  quit  |  cost  |  help");
console.log();

const rl = createInterface({ input: process.stdin, output: process.stdout });
let totalCost = 0.0;
let turn = 0;

async function ask() {
  rl.question("  You → ", async (query) => {
    query = query.trim();
    if (!query) return ask();
    if (["quit", "exit", "q"].includes(query.toLowerCase())) {
      console.log();
      console.log(`  Session ended. ${turn} queries, total cost ~$${totalCost.toFixed(6)}`);
      console.log("=".repeat(70));
      rl.close();
      return;
    }
    if (query.toLowerCase() === "cost") {
      console.log(`  💰 Total cost this session: ~$${totalCost.toFixed(6)}`);
      console.log();
      return ask();
    }
    if (query.toLowerCase() === "help") {
      console.log("  Try: 'Is payment-api healthy?'");
      console.log("       'What was deployed recently for auth-service?'");
      console.log("       'Any incidents for inventory-service?'");
      console.log("       'Give me a full readiness report for payment-api'");
      console.log();
      return ask();
    }

    turn++;
    console.log(`  ⏳ Turn ${turn} — planning and executing...`);
    const result = await runDynamicAgent(query);
    totalCost += result.cost || 0;

    console.log(`  Steps: ${result.steps}  |  Cost: ~$${(result.cost || 0).toFixed(6)}`);
    console.log("  ─────────────────────────────────────────────");
    console.log(`  ${result.report}`);
    console.log();
    ask();
  });
}

ask();
